internal/juju/applications.go

```diff
@@ -1310,6 +1323,65 @@ func (c applicationsClient) UpdateApplication(input *UpdateApplicationInput) err
 		}
 	}
 
+	// for IAAS model we process additions/removals of units
+	if modelType == model.IAAS {
+		if err = c.addUnits(input, applicationAPIClient); err != nil {
+			return err
+		}
+		if err = c.removeUnits(input, applicationAPIClient, appStatus); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c applicationsClient) addUnits(input *UpdateApplicationInput, client ApplicationAPIClient) error {
+	if len(input.AddMachines) != 0 {
+		placements := make([]*instance.Placement, len(input.AddMachines))
+		for i, machine := range input.AddMachines {
+			placement, err := instance.ParsePlacement(machine)
+			if err != nil {
+				return err
+			}
+			placements[i] = placement
+		}
+
+		_, err := client.AddUnits(apiapplication.AddUnitsParams{
+			ApplicationName: input.AppName,
+			NumUnits:        len(input.AddMachines),
+			Placement:       placements,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c applicationsClient) removeUnits(input *UpdateApplicationInput, client ApplicationAPIClient, appStatus params.ApplicationStatus) error {
+	if len(input.RemoveMachines) != 0 {
+		machineUnits := make(map[string]string)
+		for unitName, unitStatus := range appStatus.Units {
+			machineUnits[unitStatus.Machine] = unitName
+		}
+
+		unitsToDestroy := []string{}
+		for _, machine := range input.RemoveMachines {
+			unitName, ok := machineUnits[machine]
+			if !ok {
+				return fmt.Errorf("no machines deployed on machine: %v", machine)
+			}
+			unitsToDestroy = append(unitsToDestroy, unitName)
+		}
+		_, err := client.DestroyUnits(apiapplication.DestroyUnitsParams{
+			Units:          unitsToDestroy,
+			DestroyStorage: true,
```

Comment:
I'm a bit igorant on this one, but what if I had 1 unit of Postgresql and I wanted to change its placement. Would my data get lost because of this destroy storage argument?

Reply:
This is per unit storage and not storage shared for the application. I'd say you're not using best practice if you do that.